[QUEUING] [TAPE] Adding quantum tape

josh146 opened the pull request:
**Context:**

The first PR in a series of PRs, overhauling the internal structure of PennyLane.

I apologize for the size of this PR; this will be the largest PR in the series by far, as quite a few changes had to be made to lay the groundwork!

**Description of the Change:**

* `pennylane.beta.queuing`: Creates a copy of the `QueuingContext`, `AnnotatedQueue`, and `Queue` in the `beta` folder. This copy is required as the new queuing ADR requires nested queuing, which is not the current behaviour of PennyLane core. The main difference here is that only one active context (the deepest nested one) is used for appending operations. In PennyLane core, the append operation appends objects to _all_ active queues.

* `pennylane.beta.queuing.measure`: A bug was found in the previous beta measurement functions. Previously, we were using a `namedtuple` in order to represent the measurement process. Unfortunately, two measurement functions representing the same observable/return type combination results in Python simply using a reference; **not** creating a new object. For the new queuing behaviour, we need every object that is being queued to be unique.

  A solution is therefore to change the named tuple into a class.

* `pennylane.beta.queuing.operation`: Rather than duplicating all operations, observables, and Tensors to work with the new queuing approach, we can instead take advantage of the fact that the new queuing context is a context manager. Instead, we therefore:

  - Identify the existing operation/observable/tensor methods that need to be patched to support the new queuing approach.
  - generate mock methods that are monkeypatched onto the original classes when the queuing context is entered
  - remove the monkeypatched methods when exiting the queuing context.

  This allows existing operations, observables, and tensors to be used with both the current/core queuing approach, as well as the queuing approach within the `beta` folder. Once the new queuing approach is the _only_ queuing approach, the logic in the mocked methods will ported directly to the operations.

* A new quantum tape is added, that acts as a quantum operation recorder:

  ```python
  from pennylane.beta.tapes import QuantumTape
  from pennylane.beta.queuing import expval, var, sample, probs
  
  with QuantumTape() as tape:
      qml.RX(0.432, wires=0)
      qml.RY(0.543, wires=0)
      qml.CNOT(wires=[0, 'a'])
      qml.RX(0.133, wires='a')
      expval(qml.PauliZ(wires=[0]))
  ```
  
  The quantum tape is an `AnnotatedQueue`. Once the tape context is exited, the annotated queue is processed into a single linear queue:

  ```pycon
  >>> tape.operations
  [RX(0.432, wires=[0]), RY(0.543, wires=[0]), CNOT(wires=[0, 'a']), RX(0.133, wires=['a'])]
  >>> tape.observables
  [expval(PauliZ(wires=[0]))]
  >>> tape.get_parameters()
  [0.432, 0.543, 0.133]
  >>> tape.wires
  <Wires = [0, 'a']>
  >>> tape.num_params
  3
  ```

  The quantum tape is thus a core data structure representing a variational quantum circuit.

* A new circuit graph, `NewCircuitGraph`, has been added. This is currently a skeleton; it will be expanded in future PRs.

**Benefits:**

* Completes the queuing refactor

* Creates the new lightweight object that represents the variational circuit data, as specified in the queuing refactor

* Allows for nested tapes:

  ```python
  with QuantumTape() as tape:
      qml.BasisState(np.array([1, 1]), wires=[0, 'a'])
  
      with QuantumTape() as tape2:
          qml.Rot(0.543, 0.1, 0.4, wires=0)
  
      qml.CNOT(wires=[0, 'a'])
      qml.RY(0.2, wires='a')
      probs(wires=0), probs(wires='a')
  ```
  
  The nested structure is preserved:
  
  ```pycon
  >>> tape.operations
  [BasisState(array([1, 1]), wires=[0, 'a']),
  <QuantumTape: params=3, wires=<Wires = [0]>>,
  CNOT(wires=[0, 'a']),
  RY(0.2, wires=['a'])]
  ```
  
  Calling `expand` will return a tape with all nested tapes
  expanded and operations decomposed, resulting in a single tape of quantum operations:
  
  ```pycon
  >>> new_tape = tape.expand()
  >>> new_tape.operations
  [PauliX(wires=[0]),
  PauliX(wires=['a']),
  Rot(0.543, 0.1, 0.4, wires=[0]),
  CNOT(wires=[0, 'a']),
  RY(0.2, wires=['a'])]
  ```

**Possible Drawbacks:**

* Still to come: tape execution, and tape gradients.

* In all remaining docstrings, I've *left in* mentions of how the methods affect logic still to come, including execution, gradients, and interfaces. This helpfully places the method/logic in better context...

* Some components/sections of code might look weird/a bit inelegent. This is due to three reasons:

  - Optimization performed with all available components (execution/gradients/interfaces) caused us to take a route that led to less elegant code in return for faster execution. I decided not to include all components here, as it would be a 8000 line PR!

  - Backwards compatibility with existing device/operation APIs. I've left comments regarding these, and this will be cleaned up in the final PR of the series.

  - The need to support all three interfaces.

**Related GitHub Issues:** n/a

josh146 commented on pennylane/beta/queuing/queuing.py:
```diff
@@ -0,0 +1,248 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
```
This file is, for all intents and purposes, identical to the existing `pennylane._queuing` file. The only difference is the `active_context` method -- it returns only the queue on the top of the stack, not all queues on the stack. This is required to support nesting.

trbromley replied:
:+1:

Is this why there aren't any new tests written for this file? i.e., we can use `test_queueing.py` when taking this out of beta?

josh146 replied:
I will duplicate the test file as well!

josh146 replied:
done!

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -175,11 +175,11 @@ QNode is evaluated. On QNode evaluation, the quantum function is executed.
 Operators are queued to the QNode on instantiation, by having :meth:`.Operator.__init__`
 call the :meth:`.Operator.queue` method. The operators themselves queue themselves to
-the surrounding :class:`~.QueuingContext`.
+the surrounding :class:`~.pennylane.QueuingContext`.
```
Sphinx was complaining, as there are now two QueuingContext objects, two expval objects, etc.

josh146 commented on pennylane/beta/queuing/operation.py:
```diff
@@ -13,47 +13,154 @@
+    and to annotate the queue specifying previously queued objects they
+    'own'.
+    """
+    self._eigvals_cache = None
+    self.obs = []
+
+    for o in args:
+        if isinstance(o, qml.operation.Tensor):
+            self.obs.extend(o.obs)
+
+        elif isinstance(o, qml.operation.Observable):
+            self.obs.append(o)
+
+        QueuingContext.update_info(o, owner=self)
+
+    QueuingContext.append(self, owns=tuple(args))
```
Once the refactor is complete, this logic will be moved to `Tensor.__init__`, and the monkeypatching logic will be completely removed. Similarly for the other patches in this file.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+from .circuit_graph import NewCircuitGraph
+
+
+STATE_PREP_OPS = (
```
In the final PR of the series, we should move this to `pennylane/ops/qubit.py` etc. The definition of which ops are state preps should belong there, as part of the op definition.

josh146 replied:
(TBD exactly how we do this. It could be a subclass of `Operation`, or it could simply be given in the module as a module variable)

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
```
these are for backward compatibility, since devices currently expect these attributes to exist!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        self._graph = None
+        self._update_circuit_info()
+        self._update_par_info()
+        self._update_trainable_params()
```
gradient bookkeeping will also be done here in an upcoming PR

mariaschuld replied:
Just to get it right, is there a use case where the update function be called several times? Is this to have a circuit, and then change it after?

mariaschuld replied:
I think I still don't understand...if a tape is nested, and we ask for its wires, does ``tape.wires`` only have the wires visible in the highest level of nesting? And when I expand the tape, could the updating function add new wires that were hidden in a lower level? 

In other words, what metadata does a tape expansion *change*?

josh146 replied:
That's a good question.

I don't think expansion will ever change the wires/number of wires. To work out the wires on a tape, the tape loops through its queue and takes the union of all wires. So this will involve extracting the wires of any nested tapes:

```python
In [5]: with QuantumTape() as tape1:
   ...:     qml.RX(0.5, wires=0)
   ...:     qml.CNOT(wires=[0, 'a'])
   ...:     with QuantumTape() as tape2:
   ...:         qml.Rot(1,1,1,wires=[1])
   ...:         qml.CNOT(wires=[1, 'd'])
   ...:         qml.CNOT(wires=['a', 'd'])
   ...:     qml.PauliX(wires='a')
   ...:

In [6]: tape1.wires
Out[6]: <Wires = [0, 'a', 1, 'd']>

In [7]: tape2.wires
Out[7]: <Wires = [1, 'd', 'a']>
```

josh146 replied:
The reason I've kept these as separate `_update` calls, however, is I can forsee a future quantum transformation that _does_ modify the number of wires (for example, by inserting new operations that act on ancilla wires)

josh146 commented on tests/beta/queuing/test_beta_measure.py:
```diff
@@ -13,14 +13,16 @@
 # limitations under the License.
 """Unit tests for the measure module"""
```
This test file has minor modifications only:

* replacing BetaTensor with the standard PL tensor operation

* takes into account the changes with the MeasurementProcess class

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+                expval(qml.PauliZ(wires=1))
+
+    def test_observable_with_no_measurement(self):
+        """Test that an exception is raised if a measurement occurs before a operation"""
```
wrong docstring

josh146 replied:
```suggestion
        """Test that an exception is raised if an observable is used without a measurement"""
```

trbromley commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        >>> tape.measurements
+        [<pennylane.beta.queuing.measure.MeasurementProcess object at 0x7f10b2150c10>]
+        """
+        return [m[0] for m in self._obs]
```
:thinking: This is one where the difference between `measurements` and `observables` is not too clear to me

josh146 replied:
Should be fixed now!

trbromley commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+
+    @property
+    def output_dim(self):
+        """The (estimated) output dimension of the quantum tape."""
```
Is it worth elaborating on the (estimated)?

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        assert tape._graph is None
+        spy.assert_not_called()
+
+        # requested the graph creates it
```
```suggestion
        # requesting the graph creates it
```

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        # not reconstruct the graph
+        g2 = tape.graph
+        assert g2 is g
+        spy.assert_called_once()
```
:koala:

trbromley commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        rotation_gates = []
+
+        for observable in self.observables:
+            rotation_gates.extend(observable.diagonalizing_gates())
```
I remember discussing at one point making it possible for observables to be evaluated on the same wires. I guess this method is incompatible with that, but maybe that's a question for the future.

josh146 replied:
Yep. If we want to measure different observables on the same wires, we have two options:

* Simulators only: simply calculate <psi|A|psi>and <psi|B|psi>

* Hardware only: A and B must commute. Determine the rotation that rotates the circuit into the shared eigenbasis.

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        assert tape.num_params == 3
+        assert tape.get_parameters() == [params[i] for i in tape.trainable_params]
+
+        # add an additiona trainable parameter
```
```suggestion
        # add additional trainable parameters
```

trbromley replied:
```suggestion
        # add an additional trainable parameter
```

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+
+        tape.set_parameters(new_params)
+
+        for pinfo, pval in zip(tape._par_info.values(), new_params):
+            assert pinfo["op"].data[pinfo["p_idx"]] == pval
```
:grimacing:

josh146 replied:
lol. I was wondering whether to turn `pinfo` into a namedtuple? That way you could do
```python
assert pinfo.op.data[pinfo.p_idx] == pval
```

But, I'm cognizant about not making `p_info` turn back into the behemoth that `variable_deps` was. I just need it to store enough information about the parameters for tape execution/gradients

trbromley commented on tests/beta/queuing/test_beta_operation.py:
```diff
@@ -1,88 +0,0 @@
-# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
Is this to be introduced again in a later episode?

josh146 replied:
Nope; this beta tensor object has been replaced by the monkeypatching of the regular tensor.

trbromley reviewed (commented):
Thanks @josh146, looks really good! It might still be nice to discuss more in person (more for my understanding)

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        assert tape.num_params == len(params)
+        assert tape.get_parameters() == params
+
+        H2 = np.array([[0, 1], [1, 1]])
```
lol I guess this is a sneaky way to bypass any checks in `qml.Hermitian` that the matrix is actually Hermitian? :laughing:

josh146 replied:
I _think_ the checks actually happen on execution only, hence why this doesn't cause any issue!

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        assert ops[2].inverse
+
+        # check that parameter order has reversed
+        assert tape.get_parameters() == [init_state, p[1], p[2], p[3], p[0]]
```
Nice edge case you're checking here :+1:

josh146 replied:
this was a tough one to get working 😬

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+    """Tests for tape expansion"""
+
+    def test_decomposition(self):
+        """Test expanding a tape with operations that have decompositions"""
```
There are a few more that have decompositions now

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+
+        assert len(new_tape.operations) == 1
+        assert new_tape.num_params == 0
+        assert new_tape.get_parameters() == []
```
:100:

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+
+        assert len(new_tape.operations) == 3
+        assert new_tape.num_params == 3
+        assert new_tape.get_parameters() == [np.pi / 2, np.pi, np.pi / 2]
```
Why not check the operations as well? (same for above test)

josh146 replied:
done

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        assert new_tape.num_params == 2
+        assert len(new_tape.operations) == 2
+        assert isinstance(new_tape.operations[0], qml.RX)
+        assert isinstance(new_tape.operations[1], qml.RY)
```
:star_struck:

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        new_tape = tape.expand(stop_at=["Rot"])
+        assert len(new_tape.operations) == 4
+        assert "Rot" in [i.name for i in new_tape.operations]
+        assert not "U3" in [i.name for i in new_tape.operations]
```
:cookie:

josh146 replied:
🧇

co9olguy commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+            probs(wires=0), probs(wires="a")
+
+        new_tape = tape.expand(depth=2, stop_at=["PauliX"])
+        assert len(new_tape.operations) == 7
```
:champagne:

co9olguy reviewed (approved):
:sweat_smile: 

Awesome stuff @josh146 

I left a large number of comments and suggestions, but nothing is actually big enough to block an approval.

co9olguy commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        {1, 4}
+        """
+        for op in self._ops:
+            op.inverse = not op.inverse
```
This line is just flipping a boolean variable

josh146 commented on pennylane/beta/queuing/measure.py:
```diff
@@ -17,16 +17,49 @@
+from .queuing import QueuingContext
+
+
+class MeasurementProcess:
```
I got forced into it by Python's hashing behaviour lol

josh146 commented on pennylane/beta/queuing/measure.py:
```diff
@@ -17,16 +17,49 @@
+
+    def __init__(self, return_type, obs=None, wires=None):
+        self.return_type = return_type
+        self.wires = wires
```
Would it be a bit weird for a measurement process to have `wires=None`? I've imagined that all measurement processes would have to specify the wires they measure...

mariaschuld replied:
If there is a way to fill in all wires found on the tape if ``wires=None`` that would be cool. Else just make it an arg?

josh146 commented on pennylane/beta/queuing/measure.py:
```diff
@@ -53,6 +86,9 @@ def circuit(x):
         op (Observable): a quantum observable object
 
+    Returns:
+        MeasurementProcess: measurement process representing the measurement
```
Yeah, I was in two minds as well. Should we remove it from the docstring?

mariaschuld replied:
Or add it as a note for developers?

josh146 replied:
I've removed this!

josh146 commented on pennylane/beta/queuing/operation.py:
```diff
@@ -13,47 +13,154 @@
+            self.decomposition(*self.data, wires=self.wires)
         except NotImplementedError:
-            pass
+            self.__class__(*self.data, wires=self.wires)
```
Yep! The try/except is a 'hack'. Once the tape is in core, we can modify `Operation` directly and replace `decomposition` with the new `expand` method (and include sensible defaults).

josh146 replied:
I've removed this; `expand` is now recursive, and this reverts to raising a not implemented error

josh146 commented on pennylane/beta/queuing/queuing.py:
```diff
@@ -0,0 +1,248 @@
+    Queuable objects are objects that queue themselves via ``QueuingContext.append(self)``, while
+    queuing functions queue external objects.
+
+    Queuable objects can take use of the following ``QueuingContext`` methods and properties:
```
```suggestion
    Queuable objects make use of the following ``QueuingContext`` methods and properties:
```

josh146 commented on pennylane/beta/queuing/queuing.py:
```diff
@@ -0,0 +1,248 @@
+    ...     QueuingContext.append("second object")
+    ...
+    >>> print(q1.queue)
+    ['second object']
```
it depends on how the developer has written `MyQueue`. In this example, `MyQueue` is a queuing context, but not a queuable object. In the example below, as you note, `MyQueue` is now both a queuing context _and_ a queuable object.

This is just a bit of developer documentation to make that clear --- inheriting from `QueuingContext` does _not_ automatically make your object queuable itself; to make it queuable, you need to also have your object call `QueuingContext.append(self)` where necessary (like the operations do).

josh146 commented on pennylane/beta/queuing/queuing.py:
```diff
@@ -0,0 +1,248 @@
+    ...         QueuingContext.append("first object")
+    ...     QueuingContext.append("second object")
+    >>> print(q1.queue)
+    [<__main__.QueuableQueue object at 0x7f94c432b6d0>, 'second object']
```
Yep! The difference here is that the `QueuableQueue` now calls `QueuingContext.append(self)` on init.

So a bit of additional work for the developer, but I think it makes sense, and is a clean abstraction. It allows for a lot of flexibility; the developer can decide when the context queues itself. Whether this is on init, enter, or exit.

mariaschuld replied:
I think I don't get it, what is the use case of nested queues? 

I am asking because a nested tape could be quite a beautiful representation for a "batch of circuits" that can be nicely dispatched to a device by the tape. From Josh's example,

``` python 
with QuantumTape() as tape:
    qml.BasisState(np.array([1, 1]), wires=[0, 'a'])

    with QuantumTape() as tape2:
        qml.Rot(0.543, 0.1, 0.4, wires=0)
        qml.Rot(0.3, 0.0, 0.0, wires=0)
        qml.Hadamard(wires=0)

    qml.CNOT(wires=[0, 'a'])
    qml.RY(0.2, wires='a')
    probs(wires=0), probs(wires='a')
```
could be interpreted as

``` python
>>> tape.operations
[BasisState(array([1, 1]), wires=[0, 'a']),
<QuantumTape: params=6, wires=<Wires = [0]>>,
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]
```
which expands to three "paths through the tape":

``` python
>>> tape.expand()
[BasisState(array([1, 1]), wires=[0, 'a']),
Rot(0.543, 0.1, 0.4, wires=[0]),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]

[BasisState(array([1, 1]), wires=[0, 'a']),
Rot(0.3, 0.0, 0.0, wires=[0]),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]

[BasisState(array([1, 1]), wires=[0, 'a']),
Hadamard(wires=0),
CNOT(wires=[0, 'a']),
RY(0.2, wires=['a'])]
```
I am not sure how the latter would be represented, since it is a list of queues? But these three circuits constructed from different parameters and/or gates can then be sent to a device by ``tape.execute(device)``, without changing any syntax.

We could provide user UI to nest their quantum functions this way (i.e. by comparing the quantum function queues and nesting only the parts that differ), and we could have the param shift rules just be "nesting rules".

Is that a crazy thought?

josh146 replied:
The current use case of nested queues would be decompositions; for example, if a device doesn't support `Rot`, the `Rot` gate would be replaced with a queue containing the decomposition, and then expanded. So I think there is still a usecase to having all options in a nested tape simply join the parent tape.

What you suggest is cool however --- I think I have seen this in other libraries? For example, `tape.expand()` 'subsumes' nested tapes, whereas something like `tape.explode()` generate numerous tapes, by treating each operation in subtapes as a separate pathway.

I think we would still have to think of the optimum way of specifying which ops are different pathways, and which aren't

mariaschuld replied:
Can't we just a ``BatchTape`` for those that should be interpreted as parallel code, and ``explode`` only looks at the ``BatchTape`` nesting?

josh146 replied:
I like that!

co9olguy replied:
Nesting was never really considered as part of any "batching" feature, because the nested tapes all represent the *same* circuit, but at different levels of granularity. Batching would be a somewhat independent feature (e.g., using something like a list of tapes)

mariaschuld replied:
Yes, but let's say we have a batch of 1 Million quantum circuits each executed by a one-shot measurement (which appears in our current ideas around making single-shot measurements diffable). Would you really want to represent this by 1 Mio tapes and send it to the backend? Or worse, by 1 Mio Qnodes? 

If parameters _and_ gate types/templates _and_ observables are batched, for example in a VQE-like stochastic gradient descent gridsearch problem, we quickly get large batches. Wouldn't it be great to represent them as a single circuit where only the part that differs is represented by a ``BatchTape``? 

                                  StatePrep()
                                  CNOT(wires=[0,3])
                                           |
                       ---------------------------------------
                      |                    |                          |
  RY(0.1, wires=0)   RX(0.1, wires=0)   RZ(0.1, wires=0)
                      |                    |                          |
                       ---------------------------------------
                                           |
                                    CNOT(wires=[2, 1])
                                           |
                       ---------------------------------------
                      |                    |                          |
  RX(0.1, wires=0)   RX(0.2, wires=0)   RX(0.3, wires=0)
                      |                    |                          |
                       ---------------------------------------
                                           |
            -----------------------------------------------------------------------------
           |                                           |                                                |                                      
  expval(PauliZ(), shots=1)   expval(PauliZ(), shots=4)   expval(PauliX(), shots=1)

This is an exponentially compact representation which could be unfolded at will by the device before sending it off. In fact, this idea would solve so many of my current problems around shots and paralellisation with a very gentle extension of the tape structure, I would like to try a prototype if no one minds...

mariaschuld replied:
Another efficiency gain: compilers could act on the compact representation, instead of compiling 1Mio almost-copies of a circuit!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+    [[-0.45478169]]
+    """
+
+    _cast = staticmethod(np.array)
```
I think this is actually no longer used, I might remove it.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._obs = []
+        """list[tuple[~.MeasurementProcess, ~.Observable]]: measurement processes and
```
So, this is not actually ideal. Ideally, when the refactor is complete, it should work like this:

```python
self._ops = []
"""list[~.Operation]: quantum operations recorded by the tape."""

self._measurements = []
"""list[~.MeasurementProcess]: measurement processes recorded by the tape."""
```

And then, further down:

```python
@property
def observables(self):
    """Returns the observables measured by the tape"""
    return [m.obs for m in self._measurements if m is not None]
```

josh146 replied:
That is, observables are _not_ a fundamental property of the tape; measurement processes are instead.

josh146 replied:
I think I will actually make this change now, it will leave a cleaner core

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        for obj, info in self._queue.items():
+
+            if isinstance(obj, QuantumTape):
+                self._ops.append(obj)
```
Nothing will happen until later when `expand` is called 👍 

I went back and forth on this a lot, and I think this is the approach that makes the most sense. Nesting only works if the expansion is delayed, since _how_ we do the expansion depends on external variables (the device we are using, the gradient method, etc).

In the current approach, later when `tape.expand()` is called, this tape will be expanded, validation will run, and an error will be raised ('Cannot have measurements in a non-terminal position' etc.).

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+                            f"State preperation operation {obj} must occur prior to any quantum operations."
+                        )
+
+                    self._prep.append(obj)
```
Yep, this was deliberate to an extent, I didn't want to make any decisions wrt state prep validation in this PR (thought it would make sense to implement it in a future PR, when we also have a way of marking state prep operations).

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
```
This was a result of me initially trying to get tape execution to work. But I think there is a better approach, see https://github.com/PennyLaneAI/pennylane/pull/792/files#r486699359.

I might actually implement this now, it seems a lot cleaner.

josh146 replied:
Removed this, it's much cleaner and there is less code!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        self.wires = qml.wires.Wires.all_wires(
+            [op.wires for op in self.operations + self.observables]
+        )
+        self.num_wires = len(self.wires)
```
How do you mean? Do you mean simply making `self.wires` and `self.num_wires` properties, instead of attributes that are set?

co9olguy replied:
This function will get called more than once during a tape's lifetime, but these specific attributes need only be set once (they never change)

josh146 replied:
Got it!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        """Expand all operations and tapes in the processed queue.
+
+        Args:
+            stop_at (Sequence[str]): Sequence of PennyLane operation or tape names.
```
Currently, they are names e.g., `"CNOT"`. This is because, mainly:

* It works well with the existing device API; since the device API enforces all devices to return a set of strings corresponding to supported operations, you can simply do `tape.expand(stop_at=dev.operations)` for a given device.

Furthermore:

* It doesn't make sense to pass objects, because this means instantiating objects in cases where you might not already have them available (adding overhead)

* We _could_ pass the classes themselves, but this also feel clunky.

* Strings feel pythonic

co9olguy replied:
What if you want to stop at one gate, but there's another gate with the same name somewhere else at another level of nesting that you don't want to stop at? How would you differentiate?

josh146 replied:
if they are differentiated by nesting level, I would say this is best differentiated that way.

If they are at the _same_ nesting level, or you explicitly want one gate expanded but another not (regardless of nesting level)... than this is harder. One solution I can think of is if the `stop_at` list contains string _or_ objects:

```python
tape.expand(stop_at=["PauliX", "PauliY", some_op])
```

Is there a use case for this?

josh146 replied:
I've overhauled this. I realised that it is hard to guess exactly how/when a user/developer might want the stopping criterion to look.

Instead, I've made this now a callable, where you can set your own stopping condition as a function of the queue object.

E.g.:

```pycon
>>> tape.expand(stop_at=lambda obj: obj in ["PauliZ", "PauliX"])
>>> tape.expand(stop_at=lambda obj: isinstance(obj, qml.Hadamard))
>>> tape.expand(stop_at=lambda obj: obj is some_op)
```

This should be super flexible, and also similar to some other standard library functions (e.g., the `key` argument for `sorted`).

co9olguy replied:
Nice!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+                continue
+
+            t = op if isinstance(op, QuantumTape) else op.expand()
+            new_tape._prep += t._prep
+            new_tape._ops += t._ops
```
I don't think so. Currently, an `Operation` is either an operation (`RX`, `QubitUnitary`) _or_ a state prep (`QubitStateVector`, `BasisState`). I'm not sure it is possible for an operation to be both, since the distinction is made regarding the object as a whole.

The _only_ place where it becomes a bit tricky is decompositions of state preparations. For example, `BasisState` decomposes to a sequence of `PauliX` operations, which using the current definition, is no longer a state prep.

I think this motivates additional bookkeeping for marking state preps (probably better done in a separate PR). Maybe all operations have a boolean attribute that can be set as True/False, indicating whether they are part of a state prep or not?

co9olguy replied:
the presence of an op in the state prep subtape should probably be the main indicator?

josh146 replied:
By construction, I don't think could ever happen.

For example:

```python
class BasisState(operation):
    def expand(self):
        with QuantumTape() as tape:
            [qml.PauliX(w) for w, s in zip(self.wires, self.data[0]) if s == 1]
        return tape
```

The tape returned by `BasisState.expand()` will have PauliX gates in the `tape._ops` list, and nothing in the `tape._prep` list. I think we will have to somehow 'mark' operations as state preps, so that the tape knows where to put them.

co9olguy replied:
:thinking:

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+            new_tape._prep += t._prep
+            new_tape._ops += t._ops
+
+        new_tape._obs = self._obs
```
This is a good point. Diagonalizing gates are subtly different from operation decompositions:

* With operations, we can expand them out and replace them directly with the decompositions, and no information is lost.

* With observables, we can't just wholesale replace the observable (e.g., `obs=PauliX -> ops+=Hadamard, obs=PauliZ`) because we are losing information. We also need to retain information regarding post-measurement classical processing (the eigenvalues of the original observable need to be preserved).

Let me think about the best way of incorporating this. Perhaps we simply add an additional tape list attribute, to keep track of any eigenvalue post-processing that needs to be done.

co9olguy replied:
I noticed that `QuantumTape` inherits from `AnnotatedQueue` :grinning:

josh146 replied:
By the time expansion is called, however, the queue has been processed and the tape is no longer using/referring to the annotated queue!

josh146 replied:
I've added support for this. It ended up being quite elegant in the end; simply adding a `MeasurementProcess.expand()` option, and including the measurement queue in the expansion logic.

Note that devices won't be able to make use of this until we refactor their logic; for now, they will continue to access `tape.diagonalizing_gates()`.

co9olguy replied:
:fire:

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        return new_tape
+
+    def expand(self, depth=1, stop_at=None):
+        """Expand all operations in the processed to a specific depth.
```
```suggestion
        """Expand all operations in the processed queue to a specific depth.
```

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        The nested structure is preserved:
+
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         <QuantumTape: params=3, wires=<Wires = [0]>>,
```
Good idea! I think this should be an easy fix.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        old_tape = self
+
+        for _ in range(depth):
+            new_tape = old_tape._expand(stop_at=stop_at)
```
Yeah, that's been annoying me as well. Is there a good way of avoiding this?

co9olguy replied:
I feel like there's just something we haven't quite got right about the recursion. It shouldn't be done within a `for` loop, but rather truly recursively

josh146 replied:
Yeah for sure. I struggled a lot coding this up, hence why it ended up being a for loop rather than recursive. I'll see if I can re-write it to be recursive, now that I have a working version + tests.

josh146 replied:
> I've modified the logic here. expand is now recursive, so there is no need to check for fixed points/there are no useless expansions happening.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+                o1.name == o2.name for o1, o2 in zip(new_tape.operations, old_tape.operations)
+            ]:
+                # expansion has found a fixed point
+                break
```
Would it be cleaner if `Operation.expand()` raised an exception if no expansion existed, rather than simply returning itself? Perhaps that would make it easier to detect here

josh146 replied:
I've modified the logic here. `expand` is now recursive, so there is no need to check for fixed points/there are no useless expansions happening.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+
+        >>> tape.inv()
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
```
That's another solution! I can make that change.

Do you think there would ever be a use-case for wanting to invert the unitary of a tape, but preserve the state prep?

co9olguy replied:
If there is, it's an edge case, and might better be done manually

josh146 replied:
Thinking about this some more: measurements  are actually the same as state preps (they are ignored by this method). So I think we should do either:

* Tapes are only invertible if they have _no_ measurements and _no_ state preparations (operations _only_), **or**

* Inverting a tape inverts only the operations, measurements and state preps are left unchanged

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        >>> tape.inv()
+        >>> tape.operations
+        [BasisState(array([1, 1]), wires=[0, 'a']),
+         CNOT.inv(wires=[0, 'a']),
```
Not really. We'll need to modify the operations directly to fix this though (feel out of scope for this PR)

mariaschuld replied:
Would be nice to find a generic solution indeed, so all devices automatically support inversions that do not imply changing the gate itself!

josh146 replied:
Agree! This could be as simple as adding an attribute to unitary operations which are there own inverses.

Then, in `Operation.inv`:

```python
def inv(self):
    if not self.self_inverse:
        current_inv = QueuingContext.get_info(self).get("inverse", False)
        QueuingContext.update_info(self, inverse=not current_inv)
    return self
```

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        on the tape.
+
+        Args:
+            free_only (bool): if True, returns only trainable parameters
```
Oops! Tom previously suggested changing the argument name to `trainable_only`, and I forgot to do it! Will do that now

josh146 commented on tests/beta/queuing/test_beta_measure.py:
```diff
@@ -88,20 +102,24 @@ def test_annotating_tensor_hermitian(self, stat_func, return_type):
-            B = op2(1)
-            tensor_op = BetaTensor(A, B)
-            stat_func(tensor_op)
+        with contextlib.ExitStack() as stack:
+            for mock in mock_operations():
+                stack.enter_context(mock)
```
Oh, this is quite cool! So, `contextlib.ExitStack()` provides a method of entering and exiting numerous context managers at once; think of it like a context manager collection.

So instead of doing
```python
with mock1:
    with mock2:
        with mock3:
```
etc., you can instead do
```python
with contextlib.ExitStack() as stack:
    for mock in [mock1, mock2, mock3]:
        stack.enter_context(mock)
```

when you exit the `ExitStack` context, it will automate the process of exiting all contexts in the stack for you.

Here, I'm just using this to easily enter all the mock operation contexts (from `beta/queuing/operations.py`), so that the test can do `A@B` and have it correctly queue itself to the new `AnnotatedQueue`. Essentially, manually doing what the tape does in `__enter__`.

josh146 commented on tests/beta/queuing/test_beta_operation.py:
```diff
@@ -1,88 +0,0 @@
-# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
```
Good point

josh146 replied:
I've added this test back

josh146 commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+        return qml.expval(qml.PauliX(wires="a"))
+
+    # this should evaluate without error
+    func(0.432)
```
Yep! I just want to make sure that the monkeypatching of operations and tensors is removed once the tape context is exited. If this wasn't working correctly, an error would be raised here, as

* `qml.RX` and `qml.PauliX` would not find an active annotated queue to queue themselves to, and

* `func` would have no operations/observables in its internal queue to process.

josh146 commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+    def test_parameter_info(self, make_tape):
+        """Test that parameter information is correctly extracted"""
+        tape, ops, obs = make_tape
+        # tape._update_gradient_info()
+        assert tape._trainable_params == set(range(5))
+        assert tape._par_info == {
+            0: {"op": ops[0], "p_idx": 0},  # , "grad_method": "F"},
+            1: {"op": ops[1], "p_idx": 0},  # , "grad_method": "F"},
+            2: {"op": ops[1], "p_idx": 1},  # , "grad_method": "F"},
+            3: {"op": ops[1], "p_idx": 2},  # , "grad_method": "F"},
+            4: {"op": ops[3], "p_idx": 0},  # , "grad_method": "0"},
+        }
```
Yep! Once the gradient bookkeeping is introduced, each parameter will also have their grad method.

josh146 commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,667 @@
+
+        assert len(tape.queue) == 4
+        assert not tape.operations
+        assert tape._obs == [(D, C)]
+        assert tape.observables == [C]
```
Agree! Will make this change

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,777 @@
+        * ``_par_info``
+        * ``_output_dim``
+        * ``_trainable_params``
+        * ``is_sampled``
```
A quick comment: In the capabilities dictionary (where names are supposed to be *very* explicit because they can be queried by ecosystem app developers in future) I recently changed to using the term "supports_finite_shots". That was quite an aha moment, because 
a) what we really mean by sampled is that a device is run with finite shots
b) this nomenclature answers immediately the question of *what* is sampled
b) I expect after the upcoming shots ADR we decide to move all the "analytic or not" logic into the shots argument anyways

So maybe use ``finite_shots`` here?

josh146 replied:
Oh, this refers to the presence of `qml.sample()` measurements; it is independent of whether the device is run in 'analytic' mode or not. The reason we have to check for this is that the presence of sampling as output changes the output type --- we now have to return object numpy arrays.

I also don't think I can change it (at the moment), if I recall there are some assumptions elsewhere in the codebase that assume this attribute name!

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,777 @@
+
+    def _update_trainable_params(self):
+        """Set the trainable parameters"""
+        self._trainable_params = set(self._par_info)
```
Is this the one that gets overwritten by the interfaces?

josh146 replied:
yep!

mariaschuld reviewed (commented):
Great job @josh146 . I am slowly working myself into this and a lot of things are hard for me to judge as for their implications. But when you have implemented the current round of changes feel free to tag me again!

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+                # measurement process
+
+                if obj.return_type is qml.operation.Probability:
+                    self._obs.append((obj, obj))
```
tuple has been removed now :)

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        >>> tape.operations
+        [RX(0.432, wires=[0]), RY(0.543, wires=[0]), CNOT(wires=[0, 'a']), RX(0.133, wires=['a'])]
+        """
+        return self._prep + self._ops
```
Yep, this is currently just for backward compatibility --- `device.execute` expects to be able to access all prep+ops via a `.operations` attribute.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        if not free_only:
+            return params
+
+        return [p for idx, p in enumerate(params) if idx in self.trainable_params]
```
Good point. I've fixed this so we only iterate over trainable params.

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+            new_tape = old_tape._expand(stop_at=stop_at)
+
+            if len(new_tape.operations) == len(old_tape.operations) and [
+                o1.name == o2.name for o1, o2 in zip(new_tape.operations, old_tape.operations)
```
This has now been removed :)

josh146 commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,776 @@
+        >>> tape.set_parameters([4, 1, 6])
+        >>> tape.get_parameters(free_only=False)
+        [4, 1, 6]
+        """
```
I don't think so; coding the interfaces will become more complicated if there are different ways of setting all vs only trainable parameters.

I think it's just a sign of a poorly explained docstring?

josh146 commented:
@mariaschuld @trbromley @co9olguy thanks for all your suggestions! I've implemented almost everything, and it has made the PR a lot more elegant I feel.

In particular, the major ones are:

* The `tape._obs`, `tape.observables`, and `tape.measurements` attributes/properties didn't really make sense. I refactored this to make measurements a true first class citizen --- now, the internal queues are `self._prep`, `self._ops`, and `self._measurements`, and the property `self.observables` simply returns `[m.obs for m in self._measurements]`.

  This has led to much cleaner code (in fact, I was able to delete a lot of logic).

* The `MeasurementProcess` class has been made more powerful to reflect this, and to resolve a lot of the suggested changes:

  - A measurement process that involves measuring an observable can now be specified in one of two ways; by providing the observable directly, or simply by providing the eigenvalues of the observable.

  Providing the eigenvalues allows us to represent expectation values of observables 'post-rotation'; i.e., those that are measured using computational basis state measurements. Devices can access the eigenvalues for post-processing after performing the computational basis state measurements.

  - You can now call `MeasurementProcess.expand()`. This returns a quantum tape containing the diagonalization operations, and a computational basis measurement with eigenvalues for post-processing.

  - Rather than having the measurement *functions* perform the queuing, this is now done by the `MeasurementProcess` class on initialization. I noticed that they were all repeating the same queuing logic; this removes a lot of code duplication.

* The `tape.expand()` method has been completely re-written.

  - It is now _recursive_. This solves multiple problems; (a) objects with no expansion are no longer unnecessarily inspected at each depth, (b) a fixed point automatically ends the recursion, there is no need to attempt to 'check' for the fixed point manually at each loop, and (c) objects (not tapes) now raise a `NotImplementedError` if `obj.expand()` is not defined, rather than returning themselves (which was unexpected behaviour).

  - `stop_at` has been overhauled. I realised that it is hard to guess exactly how/when a user/developer might want the stopping criterion to look.

  Instead, I've made this now a callable, where you can set your own stopping condition as a function of the queue object.

  E.g.:

  ```pycon
  >>> tape.expand(stop_at=lambda obj: obj.name in ["PauliZ", "PauliX"])
  >>> tape.expand(stop_at=lambda obj: isinstance(obj, qml.Hadamard))
  >>> tape.expand(stop_at=lambda obj: obj is some_op)
  ```

  This should be super flexible, and also similar to some other standard library functions (e.g., the `key` argument for `sorted`).

  - You can now expand measurements into rotations + computational basis measurements. However, until we refactor the device logic, this won't be used during execution; devices will continue to instead access `tape.diagonalizing_gates`.

* `get_parameters()` and `inv` have been rewritten to make them significantly more efficient.

* Tests for the duplicated queuing file were added; coverage is now at 100%:

  ```
  ----------- coverage: platform linux, python 3.7.7-final-0 -----------
  Name                                          Stmts   Miss  Cover   Missing
  ---------------------------------------------------------------------------
  pennylane/beta/queuing/__init__.py                4      0   100%
  pennylane/beta/queuing/measure.py                60      0   100%
  pennylane/beta/queuing/operation.py              54      0   100%
  pennylane/beta/queuing/queuing.py                57      0   100%
  pennylane/beta/tapes/__init__.py                  3      0   100%
  pennylane/beta/tapes/circuit_graph.py            11      0   100%
  pennylane/beta/tapes/tape.py                    189      0   100%
  ---------------------------------------------------------------------------
  TOTAL                                           378      0    63%
  ```

The link to viewing the changes is available here: https://github.com/PennyLaneAI/pennylane/pull/792/commits/7a8524b129a88c2289c6ecce67cd8229e41bae36

mariaschuld commented on pennylane/beta/queuing/measure.py:
```diff
@@ -17,16 +17,157 @@
+
+        We can see that the resulting tape has the qubit unitary applied,
+        and a measurement process with no observable, but the eigenvalues
+        specified:
```
Super cool! 

So, will the ``expand()`` function always be called before sending a tape to a device? Or can the device decide? 

I'm thinking about non-qubit devices, for example homodyne measurement, where a computational basis (the Fock basis?) is very inefficient...Would the Gaussian device just ignore the eigenvalues?

josh146 replied:
It depends how we want to proceed!

* We could query the device capabilities, and use that to expand the tape in advance (like we do now with decompositions)

* We could leave the expansion to the device. This might be more elegant? But also more maintenance; changes to expansion would require updates to all the plugins

co9olguy replied:
The nice thing about having the `expand` option is it gives multiple options and doesn't force a specific one

mariaschuld commented on pennylane/beta/queuing/measure.py:
```diff
@@ -17,16 +17,157 @@
+            QueuingContext.update_info(self.obs, owner=self)
+            QueuingContext.append(self, owns=self.obs)
+        else:
+            QueuingContext.append(self)
```
Is the ``if`` for ``expval`` etc, and the ``else`` for ``probs``?

josh146 replied:
yep!

mariaschuld commented on pennylane/beta/queuing/measure.py:
```diff
@@ -140,11 +273,7 @@ def circuit(x):
-    qml.QueuingContext.append(meas_op, owns=op)
-
-    return op
+    return MeasurementProcess(Sample, obs=op)
```
Great to avoid the duplication!

mariaschuld commented on pennylane/beta/queuing/measure.py:
```diff
@@ -182,11 +311,4 @@ def circuit():
-    qml.QueuingContext.append(meas_op, owns=op)
-
-    return op
+    return MeasurementProcess(Probability, wires=qml.wires.Wires(wires))
```
Just as a thought...we spoke about dummy measurements before. This could be great for debugging, or when one wants to access the state of the device. Could this be done with an empty  ``MeasurementProcess``?

josh146 replied:
Yes, I was thinking about that as well. The only issue is that currently devices won't know what to do with a measurement process that isn't a expval/prob/var/sample! So something to consider in the device refactor

mariaschuld commented on pennylane/beta/queuing/operation.py:
```diff
@@ -13,47 +13,154 @@
+    # used as arguments.
+    mocks += [mock.patch.object(qml.operation.Operator, "do_check_domain", False)]
+
+    return mocks
```
Yes, is it just the old operations + some queuing monkeypatching? Or does anything change?

josh146 replied:
Within the `Tape` context, this changes the following methods:

* `Operation.queue`: to force operations to queue themselves to the beta queue

* `Operation.inv`: to force inverted operations to annotate the queue with their inversion, rather than changing an internal attribute. This allows the _same_ operation object to be used in different queues with different inverse status (which makes more sense! There have been weird edge cases where inverting an operation in one qnode causes it to be inverted _everwhere_).

* `Operation.expand`: since this doesn't previously exist. This is just a wrapper around `Operation.decomposition`, but it returns a tape rather than a list.

* `Tensor.init`: Currently, tensors don't perform any queuing, so we need to make sure tensors both (a) queue themselves, and (b) annotate the queue declaring any previous observables that they 'own'.

* `Tensor.__matmul__`, `Tensor.__rmatmul__`: we need to also patch these as per above.

Once the tape context is exited, all of these patches are removed.

When the tape is moved to core proper (and the old core removed), these 'patches' should be moved to become officially part of the operations.

josh146 replied:
But happy to discuss 'in person' as well!

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,835 @@
+            new_tape._ops += expanded_tape._ops
+            new_tape._measurements += expanded_tape._measurements
+
+    return new_tape
```
Nice. I remember when learning F# I thought quite often that recursion is the single most underutilised programming paradigm in python!

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,835 @@
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._measurements = []
+        """list[~.MeasurementProcess]: measurement processes recorded by the tape."""
```
Just a thought: With this design we will never be able to support intermediate measurements, no? Which is fine because they can always be moved to the end. But what happens in the case of intermediate measurements that condition future gates on measurment results? Can those also be moved to the end? I'm thinking of a measurement-based QC design...

josh146 replied:
Yep I had this same thought as well. There are so many assumptions everywhere regarding measurements being at the end of the circuit.... I didn't know whether to just stick with it for now (less work) or change it

josh146 replied:
Perhaps we could add tapes that support intermediate measurements at some points... allows us to keep both behaviour for a while, and avoids massive refactors

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,835 @@
+                if self._measurements:
+                    raise ValueError(
+                        f"Quantum operation {obj} must occur prior to any measurements."
+                    )
```
When could this happen?

josh146 replied:
e.g.,

```python
with QuantumTape() as tape:
    expval(qml.PauliZ(0))
    qml.RX(0.56, wires=1)
```

It essentially forces all operations to precede all measurements in the queue (even if an operation and measurement are 'independent' as above). This is the same restriction we have currently with the QNode, but we never really noticed it since the `return measurements` pattern was so entrenched.

flatslm replied:
Hi
I try use:
@qml.qnode(dev)
def circuit1(N: int):
  qml.PauliX(0)
  w = range(0, 1)
  result = qml.expval(qml.PauliX(0))
  qml.Hadamard(wires=[1])
  return result

but got:
ValueError: Quantum operation Hadamard(wires=[1]) must occur prior to any measurements.

qml.version() = 0.20.0

It it an old issue or it need create a new one?
Thanks
@flatslm

josh146 replied:
Hi @flatslm! This occurs because `qml.expval()` must always be the last instruction in the QNode. For example, the following should work:

```python
@qml.qnode(dev)
def circuit1(N: int):
    qml.PauliX(0)
    qml.Hadamard(wires=[1])
    return qml.expval(qml.PauliX(0))
```

For future reference, I suggest posting questions on the [PennyLane discussion forum ](https://discuss.pennylane.ai/) rather than GitHub PRs, as they will be much more visible 🙂

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,835 @@
+
+            if queue == self._ops:
+                # reverse the list representing operator parameters
+                obj_params = obj_params[::-1]
```
Later one could change the parameters whose inverse gate just take ``(-1) param`` here...

josh146 replied:
Oh nice idea!

We would have to have a way for operations to declare _how_ they are invertible; it could be either a

* linear combination of parameters (includes the case of `-param[0], param[1], param[2]`)

* something more complicated (e.g., `param[0].conj().T` for `QubitUnitary`)

* no parameters; cannot be inverted this way, relies on the device implementation.

mariaschuld commented on pennylane/beta/tapes/tape.py:
```diff
@@ -0,0 +1,835 @@
+
+    # ========================================================
+    # properties, setters, and getters
+    # ========================================================
```
isn't the ``trainable_params`` function above also a setter?

josh146 replied:
Good point, I'll change this

mariaschuld commented on tests/beta/queuing/test_beta_measure.py:
```diff
@@ -129,17 +140,113 @@ class TestBetaProbs:
+        assert len(tape.operations) == 4
+
+        assert tape.operations[0].name == "Hadamard"
+        assert tape.operations[0].wires.tolist() == [0]
```
Just a comment, also possible to use ``wires.labels == (0,)``

josh146 replied:
Oh of course 🤦 I was so used to seeing `.tolist()` during the wires refactor, it was stuck in my head!

mariaschuld commented on tests/beta/tapes/test_tape.py:
```diff
@@ -0,0 +1,711 @@
+            ops += [qml.RX(0.133, wires=4)]
+            obs += [qml.PauliX(wires="a")]
+            expval(obs[0])
+            obs += [probs(wires=[0, "a"])]
```
How come you are constructing an invalid tape here (since you measure the same wire three times)? Is that because it is never checked by the tape class itself, and convenient?

josh146 replied:
I have to admit, I haven't added a validation here regarding measuring wires more than once... I was thinking we could come up with a better solution down the line

mariaschuld reviewed (approved):
I read through a second time, and although I still cannot judge 100% how this PR fits into the larger picture it looks sound. Very elegant code and extremely useful examples in the docstring!

trbromley commented on pennylane/beta/queuing/measure.py:
```diff
@@ -86,9 +194,6 @@ def circuit(x):
         op (Observable): a quantum observable object
 
-    Returns:
-        MeasurementProcess: measurement process representing the measurement
```
Do we still return a `MeasurementProcess` but just don't want to highlight?

josh146 replied:
Yep

trbromley reviewed (commented):
Thanks @josh146!

trbromley reviewed (commented):
Thanks @josh146!

trbromley commented:
Is the codecov report accurate? 
![image](https://user-images.githubusercontent.com/49409390/93084446-09ced800-f662-11ea-8a0a-cb27e6dd578f.png)

josh146 commented:
@trbromley, I think this is codecov playing up 🤔  This is what I get locally:

![image](https://user-images.githubusercontent.com/2959003/93085870-66b7a600-f6d5-11ea-8111-27c28c4edb52.png)

co9olguy commented on pennylane/beta/queuing/measure.py:
```diff
@@ -61,6 +78,97 @@ def __init__(self, return_type, obs=None, wires=None):
+
+        **Example:**
+
+        >>> m = MeasurementProcess(Expectation, obs=qml.PauliX(wires=1))
+        >>> U.eigvals
+        >>> array([1, -1])
```
Looks like there are typos in this docstring @josh146

josh146 replied:
🤕

josh146 replied:
Will fix it in the latest PR. Thanks!

co9olguy commented on pennylane/beta/queuing/measure.py:
```diff
@@ -61,6 +78,97 @@ def __init__(self, return_type, obs=None, wires=None):
+        rotation and a measurement in the computational basis.
+
+        Returns:
+            .QuantumTape: a quantum tape containing the operations
```
Just want to confirm, the `.` here resolves correctly in built docs? (no ~ needed?)

co9olguy commented on pennylane/beta/queuing/measure.py:
```diff
@@ -61,6 +78,97 @@ def __init__(self, return_type, obs=None, wires=None):
+
+        Returns:
+            .QuantumTape: a quantum tape containing the operations
+            required to diagonalize the observable
```
indent?

josh146 replied:
Nope, the returns statement can only have a single value, so there is no indent on line continuation (unlike `Args:`). Including the indent here will cause rendering issues.
